jgrapht-core/src/main/java/org/jgrapht/alg/clustering/LabelPropagationClustering.java

```diff
@@ -0,0 +1,360 @@
+            result =
+                new ClusteringImpl<>(new Implementation<>(graph, rng, maxIterations).compute());
+        }
+        return result;
+    }
+
+    /**
+     * The actual implementation
+     *
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    private static class Implementation<V, E>
+    {
+        private Graph<V, E> graph;
+        private Random rng;
+        private int maxIterations;
+        private Map<V, String> labels;
+
+        /**
+         * Initialize the computation
+         * 
+         * @param graph the graph
+         * @param rng the random number generator
+         * @param maxIterations maximum iterations
+         */
+        public Implementation(Graph<V, E> graph, Random rng, int maxIterations)
+        {
+            this.graph = graph;
+            this.rng = rng;
+            this.maxIterations = maxIterations;
+            this.labels = new HashMap<>();
+
+            int i = 0;
+            for (V v : graph.vertexSet()) {
+                labels.put(v, String.valueOf(i++));
+            }
+        }
+
+        /**
+         * Main loop of the algorithm
+         * 
+         * @return the clusters
+         */
+        public List<Set<V>> compute()
+        {
+            int currentIteration = 0;
+            while (true) {
+                // is there a limit on the number of iterations?
+                if (maxIterations > 0 && currentIteration > maxIterations) {
+                    break;
+                }
+
+                // perform synchronous label update (to avoid oscillations)
+                boolean anyChange = false;
+                List<V> allVertices = new ArrayList<>(graph.vertexSet());
+                Collections.shuffle(allVertices, rng);
+                for (V v : allVertices) {
+                    if (updateLabel(v)) {
+                        anyChange = true;
+                    }
+                }
+
+                // stopping criterion
+                if (anyChange == false || shouldStop()) {
+                    break;
+                }
+
+                currentIteration++;
+            }
+
+            return computeCommunities();
+        }
+
+        /**
+         * Stopping criterion. Perform the iterative process until every node in the network has a
+         * label equal to a label that the maximum number of its neighbors belong to.
+         * 
+         * @return true whether we should stop, false otherwise
+         */
+        private boolean shouldStop()
+        {
+            for (V v : graph.vertexSet()) {
+                Pair<Map<String, Integer>, Integer> labelCountsAndMaximum =
+                    getNeighborLabelCountsAndMaximum(v);
+                Map<String, Integer> counts = labelCountsAndMaximum.getFirst();
+
+                String vLabel = labels.get(v);
+                int vLabelCount = counts.getOrDefault(vLabel, 0);
+                for (E e : graph.edgesOf(v)) {
```

Comment:
As far as I understand, what we are trying to do here is to find a neighbour of `v` who's label count is higher than the label count of `v`. We can avoid traversing the neighbour edges of `v` again by traversing the `counts` map entries instead. As an alternative, it's possible to compute the desired maximum label count max value immediately while computing the `counts` map in a helper function. 

I've tested the performance difference and it was really marginal (~ 3-5% for graphs of 1000/10000 vertices generated using Kleinberg graph generator), so I'm not sure if it makes sense to apply these changes.

Reply:
Yes you are right. If you have the changes somewhere, feel free to commit on this branch.